src/models/graph.ts

```diff
@@ -0,0 +1,457 @@
+
+export type GraphConnectionProps = {
+	sourceNodeId: string;
+	sourcePortId: string;
+	sinkNodeId: string;
+	sinkPortId: string;
+	type: ConnectionType;
+}
+
+export class GraphConnectionRecord extends ImmuRecord<GraphConnectionProps>({
+
+	sourceNodeId: "",
+	sourcePortId: "",
+
+	sinkNodeId: "",
+	sinkPortId: "",
+
+	type: ConnectionType.Audio
+
+}) {
+
+	public get id(): string {
+		return GraphConnectionRecord.idFromNodesAndPorts(this.sourceNodeId, this.sourcePortId, this.sinkNodeId, this.sinkPortId);
+	}
+
+	private static readonly connectionDelimiter = "__=__";
+	public static idFromNodesAndPorts(sourceId: string, sourcePortId: string, sinkId: string, sinkPortId: string): string {
+		return `${sourceId}:${sourcePortId}${this.connectionDelimiter}${sinkId}:${sinkPortId}`;
+	}
+
+	public static connectionsFromDescription(nodeId: GraphNodeRecord["id"], desc: OSCQueryRNBOInstanceConnections): GraphConnectionRecord[] {
+		const conns: GraphConnectionRecord[] = [];
+
+		// Node as source
+		for (const [portId, info] of Object.entries(desc.CONTENTS.audio?.CONTENTS?.sources?.CONTENTS || {})) {
+			const commonConnProps: Omit<GraphConnectionProps, "id" | "sinkNodeId" | "sinkPortId"> = {
+				sourceNodeId: nodeId,
+				sourcePortId: portId,
+				type: ConnectionType.Audio
+			};
+
+			conns.push(...(info.VALUE.map(target => {
+				const [sinkNodeId, sinkPortId] = target.split(":");
+				return new GraphConnectionRecord({
+					...commonConnProps,
+					sinkNodeId: sinkNodeId === GraphSystemNodeRecord.systemName ? GraphSystemNodeRecord.systemOutputName : sinkNodeId,
+					sinkPortId
+				});
+			})));
+		}
+
+		for (const [portId, info] of Object.entries(desc.CONTENTS.midi?.CONTENTS?.sources?.CONTENTS || {})) {
+			const commonConnProps: Omit<GraphConnectionProps, "id" | "sinkNodeId" | "sinkPortId"> = {
+				sourceNodeId: nodeId,
+				sourcePortId: portId,
+				type: ConnectionType.MIDI
+			};
+
+			conns.push(...(info.VALUE.map(target => {
+				const [sinkNodeId, sinkPortId] = target.split(":");
+				return new GraphConnectionRecord({
+					...commonConnProps,
+					sinkNodeId: sinkNodeId === GraphSystemNodeRecord.systemName ? GraphSystemNodeRecord.systemOutputName : sinkNodeId,
+					sinkPortId
+				});
+			})));
+		}
+
+		// Node as sink gets only applied for system node connections
+		for (const [portId, info] of Object.entries(desc.CONTENTS.audio?.CONTENTS?.sinks?.CONTENTS || {})) {
+			const commonConnProps: Omit<GraphConnectionProps, "id" | "sourceNodeId" | "sourcePortId"> = {
+				sinkNodeId: nodeId,
+				sinkPortId: portId,
+				type: ConnectionType.Audio
+			};
+
+			for (const target of info.VALUE) {
+				const [soureNodeId, sourcePortId] = target.split(":");
+				if (soureNodeId !== GraphSystemNodeRecord.systemName) continue;
+				conns.push(new GraphConnectionRecord({
+					...commonConnProps,
+					sourceNodeId: GraphSystemNodeRecord.systemInputName,
+					sourcePortId
+				}));
+			}
+		}
```

Comment:
we can create RNBO nodes that are sink only:

<pre><code>
----------begin_max5_patcher----------
3059.3oc6bk0baiiD9Y6eEn3KY2prUH3gj37zjMwaUdqMwoxw7RpoTAQBIyY
oH0RBZKOSM4291.fG.fGRNVI6Xm7PhswQyF8G5Fc2rI9iSOwZY1NZgE5mPeB
cxI+wombhnIdCmT82mXsgrKLgTHFlUJ81rk+l0YxttgjmR1PE8jmtL6y0cvn
6XcaMsbSVIKgxDzBW0prI1caoRFwJIt.l7uV0cbjfPvS8bbMg1RXgWGmtdQN
MjImEFa6Mw9Lji+T9O7rE+gyD6FBAO83TyGNm+tglWDmkJdL3ItSrOOhdi1i
hlWIRpjImXsJNgpLuJxchEY6VklOQYJbA4ukIHTvYMMEmJaxtoob5Mw0yusU
RNrfYvpsLWJt2M0ypkLYQz7zxXAqHa7OOslkDfGGlJ1RBawp5Ya0JC8mgERs
4BIIdZf3myCrakhv9i0IYg+GZjJ+YksklFmtMmVPSYDlA6aEQWQJSXKVkkxJ
h+cASfArou9WUwj81Yyts+Mg0tBVmGGkkx4AMrf2b8SC1f3KVT9pqEwHRIa6
Yxv9MPtLPmEvZrrXIImCUKSDOAm5NYYYI5c0LuD5JVU2aiSSMDhrrsC2Yd75
qGYtKyfN2LFsE8TrnLU16BXWAaQA4FcgMijjToiqS9cjz3MDFkEKQ.G6lNoo
DXgdcQXdVRh15U1yM8zCngEGRuMNhcs3AotW.Fd7158PVMnbT7ZZASuMFYcg
dKEr6jBcklJWVoEufQ2rMAVE5C.zO.SNEWmcaQ0.q2m0JWhYUjsLU76QVJxl
VanpJ7p1R0ZeLapcrqdCIQsqFKqaI4jMHnWzOCVQPmigeR1gvpCV2fqiROcL
5ZcF7uFMCCCutpDsWiuNBsKOuoRkLocDWEcsNVfU4Ft8n3zUYJBKEoELQJMp
3xTPsKMjdY0.wmoLBR9ZgINUJnRBtLMojp0sV+.IXr7qxead1VC1yXy26oLT
bZLKFD8RhZLTo8Jfi5gNwEjjXRglw8pGOu85MRphMwVX.qHFVSZYaR300cZa
93DaS5omHJXsNDTEh5oy58EbDaIb3WuKveoVfZYaLfDxRpPc25xJ4zuzmbBX
tDxcY4QxiWcU5s4zqJ7OKeCHe9cZz8DEwihhBhgZI9Dj0QAs9Z.IBEmdVRlP
0vBQ9gyrign6cbBgj1eXYipKbTDf3i6d5kjz0VGxdsigr5MMTSJvnLZtTRM4
ut60jGN7f1rUe70C0X6a.5fxVgXWSaEfiZvE+n1fawcaVlkLn8z21rGhKXLF
1FRZD3Vb9cJgAzG1PAz6XfMerfhHoHN4n47EFB13rsjY8nFADQeNj7+BXsJO
LqXOmlMcX4O3DT7liCD7ZIodH1d4ri0SE+MpjG6AavihMeCwEbO3xUqDTXSu
qjmjNI9vAMxtilBkjTOLEJxtGYJT3gwlJQ63XiynXy2PbYLEpdWIOhUn9JBZ
zcayRooriR3xgjDpD2JP0DFhWH4NquCBMdXP5hZY73nzrgQoBFcaww.hdU7M
wQTgO1RO3fX5f.6HH4Zj6+MvVg4fimH4C865jZ79dDAlv17ggspg9liTDRu5
h29tKd4K9vEu5mPbOx4nXR1s.rEBhdzypdb7.xdFfq.9Qhdbfe3urXjdkbAK
iPZuvvwJP0Wf1jkSQ4fzkurQb5hVkkqG4J..hPl1w34XNA8t27OtBwibk9HO
podwjN5TGFlsG+.8FFSKSiYGGvTtd3o2R99HVJUr3OfNoi3GfZ+f5G4nwd.S
+gASV1WRp3FFTeYV5MzbVAbtFnklXlONQpLgdaenb02MSrdJmOoOjgTxPI3R
RNsP7xm2CrYOLrsJOayWMfKsuzoJbVgCcjPVYWf8oMB9OAw8WDFFLLD1LlGJ
xcEmPbyi2dcb30Muo.C.p.cabRBBLwxqd.tKnnrTT0ahE7kgDMA8AtuoBpEB
lZggJR5XbHzXksZ32+aOaUbdA6Yf03Hzy3uf6m82+dvB7vtp9N.I3tkH.h8o
T6rWmk9ZswnYCQ0VA9qhGUtUbvJhmPV9YtPHiJC7GaIFcKw460UqCZKwHuSV
PWMJ9H4u06AZ8038H9+KLZY19QHVdmUoRHlbAxk62+oQ7ENjkbjxk8ku5RTX
VJiWDOf1VURAfybqZD7FNtnU0bxSvrCnjaMtz3ksRi2HF9j8oHM1qChxHGmf
V9Wu+p2HbZElB3JPAKONcs.kHfxE7q7GUDgQDAlVVPQKuSDJCOsb47oDlEQ+
tNlkWWIg1GdNarBHX68rTKFI4NYfKuLDIJRpj0DohH.TwiA8o1L.gNG8w2eA
5mq5wtNeO+5jG2Y74PrmthjTLV9dNdfxKFAMl7c9AWWlFlTFQgytPuUJv2ih
j9q06z9QOq3zskLshrrK7oVwjc3v8j72Vv8s5wlXLvqy5qrXhZKHZ88e6mWM
qrsuPt9fplsgXSsMeVx2.vSaY8WJCe3Uy0gIqullrsg8jZtZE0pxqurauM00
+.8mskWUNFedAhJgU9AJbof8NGG3YGLCaG3oQ7qIovx9sxfvu7J8Uhpxon3c
WP2wxIK3VPiWVxzqCW0J4s0JrV46ZjLdsymaqb21pDuyaoDq2Uchf0ITiKwF
DxLSi5yxzjkdrU8juK8oWWsU5s1RMsf2aemdFrnZzVZRtZuGsrFAc.ww+sjV
UP2FETtX.EPfyjDChK6pusYmZZvtU+6gW76MU3N+iun.rB7Yji8vU0NdzpZe
nJZ24.qncWYEsO299WP6GrNA+q2oBxOTLrQzrnyh49gmMD5aDlFm94w9DEFE
LKhWmxWTC.o3CDRclK95Db82OjZ+M9aT.Vx3iQUGKkTHtoIjX0ftMlwy+ZDU
+KD4H4hp8eEpI4psGCGAfX4eLB.PJRGqLBdRTKx8t.OzhLNLayliTEDAFHJn
H9moIplpOERQwOxHzSoLB4duii8gD1k3fhwsCdbNbvnXkrd3g4.d.LXPNl8o
6qhYuGZ.NylZaOaFFuu3avGg3aTL6o6v+82OcX4tnieM2Ou63Rrd7qq8CfFh
GZnucUgGT7962guhrx7vZOypbBEo6JUDriKNs4qA+SMeKo7wcPtade4A7AxC
NCxCshl5O67kqCyRjax9D2EcWr6LGeu.abvT64.clfa9yYy8bw7lbvvtNuY1
9ym6356dFO9zVRqQv.m.OryYZ+lqO+WMlEMgx2ZoMYW+Y9d1yblNcJd5bwb8
w99S8CBvd13oyDMMeJd17oyl6LyC66MyftjvvNj0y010Cim6gcCb7Ej0dpiq
suqKF9AdpjrNN191ymicm6O2fp7iZ1R0npi+bW241ScC78csmUsP8b.Ndtsi
6bWOI+iCvAdAy87bmZ6YP1kqWEmjHH6hFKdxmReC4nCccHN9qJ0cZjcScqvW
aaOeOAsrcmOKX1Hj.LwUc2DLS4CrWcDaAeHfy4quuFl3FzKcJYYqyIQwUV1p
CTs1.lzg3KUBNp9Zx3xqpasqAlAOvp0Q8Vya7uz+JSZJ1C0M2pnDqydxXG1G
+YNKt8TtA+Pv0Dp4DuQZ+UaIo3zfZNV.GXtnJGPs2mGfKMfiRB2wTY9JKQM7
.WxWeGMTekkTGxdeWIIMxzWNy680iqFRUtXRZtrJDW.IJczrKoSFBT+t3rxx
iWKSkl1M7x.eFc7afglN9H.tqhSUt.EnPXFQ52uFFKwVQrpbWYujZJBO07zU
qEKjocr5f0kNgT+UAQmOKvd04XLkdd.+OIgQTu411Xriik4snhlXnCaEVlmC
6CWnx3sfMzSdr94rpGisOfa.raL3SCAMuwJL.vCRbnlKmZVU8VjwDkUauAeU
4tNPrgiTZ.8ICcWNnlxLCWs+xAcSWy33SNc0fI9YD4rbxcFwBG37NL1FOYCY
GWRzYJfmOhKFEqO+7WkEVxMxU77WS1gBdd81ihtSJqP5wqVliqFNe.UxhYvg
69Ki7b3GP4EszltzwIXUjCMHbkM0VUNb5IFVU6XihaeJZg79xoGOjqMO07BS
VHuUXzT3KKu2vjwyu4AOBKH11jspq5aKuEm1y81i5.RxRW26cdk1n.AdNa+C
qVcw0PxdZ8+W4Pp40JEmwMtNoLtJo5dMRM7UHk40GkPx120FEO6um9mm9+L2
93lN
-----------end_max5_patcher-----------
</code></pre>

this doesn't show up in the graph but loads on the runner just fine

Reply:
should be supported now